app/src/chat_api.py

```diff
@@ -0,0 +1,240 @@
+    response_text: str
+    citations: list[Citation]
+
+    # Populated after instantiation based on LiteralAI message ID
+    response_id: Optional[str] = None
+
+
+def get_chat_engine(session: UserSession) -> ChatEngineInterface:
+    engine_id = session.chat_engine_settings.engine_id
+    # May want to cache engine instances rather than creating them for each request
+    engine = (
+        chat_engine.create_engine(engine_id) if engine_id in session.user.allowed_engines else None
+    )
+    if not engine:
+        raise HTTPException(status_code=406, detail=f"Unknown engine: {engine_id}")
+    for setting_name in engine.user_settings:
+        if setting_value := getattr(session.chat_engine_settings, setting_name, None):
+            setattr(engine, setting_name, setting_value)
+    return engine
+
+
+@router.post("/query")
+async def query(request: QueryRequest) -> QueryResponse:
+    # For now, use the required session_id as the user_id to get a UserSession
+    session = await _get_user_session(request.session_id)
+    with literalai().thread(name="API:/query", participant_id=session.literalai_user_id):
+        request_msg = literalai().message(
+            content=request.message,
+            type="user_message",
+            name=request.session_id,
+            metadata={
+                "request": request.__dict__,
+                "user": session.user.__dict__,
+            },
+        )
+
+        # May want to cache engine instances rather than creating them for each request
+        engine = get_chat_engine(session)
+        response: QueryResponse = await run_query(engine, request.message)
+
+        # Example of using parent_id to have a hierarchy of messages in Literal AI
+        response_msg = literalai().message(
+            content=response.response_text,
+            type="assistant_message",
+            parent_id=request_msg.id,
+            metadata={"citations": [c.__dict__ for c in response.citations]},
+        )
+        # id needed to later provide feedback on this message in LiteralAI
+        response.response_id = response_msg.id
+    return response
+
+
+async def run_query(engine: ChatEngineInterface, question: str) -> QueryResponse:
+    logger.info("Received: %s", question)
+    chat_history = None
+    result = await asyncify(lambda: engine.on_message(question, chat_history))()
+    logger.info("Response: %s", result.response)
+
+    final_result = simplify_citation_numbers(result)
+    citations = [Citation.from_subsection(subsection) for subsection in final_result.subsections]
+    return QueryResponse(response_text=final_result.response, citations=citations)
+
+
+# endregion
+
+logger.info("Chat API loaded with routes: %s", router.routes)
+
+
+def main() -> None:  # pragma: no cover
+    import getopt
+    import sys
+
+    import uvicorn
+    from fastapi import FastAPI
+
+    # Use default port 8001 so that this standalone API app does not conflict with the Chainlit app
+    port = 8001
```

Comment:
What's the reason to run this on a different port when the API is invoked standalone? (And as a result, if I'm following right, only support standalone mode locally and not in the Docker container?)

Reply:
I intend this for `Only for local development` currently. If/once we decide to run the API without the Chatbot web UI, then we can make this a more permanent change and use the application port 8000.

Reply:
Right, my question is why -- it seems like if you just leave it on port 8000 it'll work in Docker too, right?

Reply:
Yes, I think we should continue to support both methods of running the app locally: either directly on the machine (poetry run...) or in a Docker container.